here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/AbstractTask.java

```diff
@@ -624,23 +621,51 @@ public final boolean removeListener(@NotNull Consumer<@NotNull RESULT> listener)
     }
   }
   /**
-   * Increments the value of authorUsage in the hashmap.
+   * Increments the value of authorUsage in the hashmap atomically.
    *
-   * @param author The author.
+   * @param principal The author.
    */
-  private static void incrementAuthorUsage(String author) {
-    authorUsageMap.computeIfAbsent(author, key -> new AtomicInteger()).incrementAndGet();
+  private static void incrementPrincipalUsage(String principal) {
+    if (principal != null) {
+      AtomicInteger authorUsageValue =
+          principalUsageMap.merge(principal, new AtomicInteger(0), (existingValue, newValue) -> {
+            existingValue.incrementAndGet();
+            return existingValue;
+          });
+    }
   }
 
   /**
-   * Decrements the value of authorUsage in the hashmap.
+   * Decrements the value of authorUsage in the hashmap atomically.
    *
-   * @param author The author.
+   * @param principal The author.
    */
-  private static void decrementAuthorUsage(String author) {
-    AtomicInteger usage = authorUsageMap.get(author);
-    if (usage != null) {
-      usage.decrementAndGet();
+  private static void decrementPrincipalUsage(String principal) {
+    if (principal != null) {
+      AtomicInteger authorUsageValue = principalUsageMap.computeIfPresent(principal, (key, oldValue) -> {
+        if (oldValue.get() > 0) {
+          oldValue.decrementAndGet();
+        }
+        return oldValue;
+      });
     }
   }
+
+  /**
+   * Returns the  value of author name.
+   */
+  private String getAuthorName() {
+    return context.getAuthor();
+  }
+
+  /**
+   * Returns the  value of principal name.
+   */
+  private String getPrincipalName() {
+    String author = context.getAuthor();
+    //TODO uncomment below two lines after fixing unit tests.
+    // String appId = context.getAppId();
+    // return author + appId;
```

Comment:
use concatenation of "appId" + "#" + "author" to create value for principal.

Reply:
We decide to consider new logic for this so not required now.